src/dodal/common/beamlines/instantiation_behaviour.py

```diff
@@ -0,0 +1,131 @@
+    The device is then cached here locally, and also in the global variable ACTIVE_DEVICES.
+    """
+
+    device: AnyDevice | None = None
+    factory: Callable[[], AnyDevice]
+
+    def __init__(
+        self, config: DeviceInitializationConfig, factory: Callable[[], AnyDevice]
+    ) -> None:
+        self.factory = factory
+        self.config = config
+        CONTROLLERS[self.factory.__name__] = self
+        self.device = None
+
+    def see_if_device_is_in_cache(self) -> AnyDevice | None:
+        return self.device
+
+    def __call__(
+        self,
+        connect=False,
+        mock: bool | None = None,
+        timeout: float | None = None,
+        name: str | None = None,
+    ) -> AnyDevice:
+        device = self.see_if_device_is_in_cache() or self.factory()
+        if self.config.set_name:
+            device.set_name(self.factory.__name__)
+        self.add_device_to_cache(device)
+
+        if connect:
+            call_in_bluesky_event_loop(
+                device.connect(
+                    timeout=self.config.default_timeout_for_connect
+                    if timeout is None
+                    else timeout,
+                    mock=self.config.default_use_mock_at_connection
+                    if mock is None
+                    else mock,
+                )
+            )
+            assert self.device is not None
+            CONTROLLERS[self.device.name] = self
+
+        if self.device is not None:
+            CONTROLLERS[self.device.name] = self
+        return device
+
+    def add_device_to_cache(self, device: AnyDevice) -> None:
+        self.device = device
+        ACTIVE_DEVICES[device.name] = device
+
+
+CONTROLLERS: dict[str, DeviceInitializationController] = {}
+
+
+def instantiation_behaviour(
+    *,
+    eager: bool = True,
+    set_name: bool = True,
+    default_timeout_for_connect: float = DEFAULT_TIMEOUT,
+    default_use_mock_at_connection: bool = False,
+) -> Callable[[Callable[[], AnyDevice]], DeviceInitializationController]:
+    config = DeviceInitializationConfig(
+        eager, set_name, default_timeout_for_connect, default_use_mock_at_connection
+    )
+
+    def decorator(factory: Callable[[], AnyDevice]) -> DeviceInitializationController:
+        controller = DeviceInitializationController(config, factory)
+        return controller
+
+    return decorator
+
+
+beamline_prefix = "example:"
+
+
+@instantiation_behaviour(
+    default_use_mock_at_connection=True, default_timeout_for_connect=10
+)
+def new_detector_xyz():
+    """Create an XYZ detector with specific settings."""
+    return XYZDetector(name="det1", prefix=f"{beamline_prefix}xyz:")
+
+
+@instantiation_behaviour(
+    eager=True, default_use_mock_at_connection=True, default_timeout_for_connect=10
+)
+def detector_xyz_variant():
+    """Create an XYZ detector with specific settings."""
+    return XYZDetector(name="det2-variant", prefix=f"{beamline_prefix}xyz:")
```

Comment:
Could: Could the example module be in its own file? I'd find that more readable.

Reply:
sure! and path and name suggestions @callumforrester ?